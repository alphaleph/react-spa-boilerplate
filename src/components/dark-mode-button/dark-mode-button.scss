.is-dark {
    background-color: black;
    color: white;
}

.is-light {
    background-color: azure;
}