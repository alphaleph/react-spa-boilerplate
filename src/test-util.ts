import { BaseAction } from './types';

export const defaultDispatch = (action: BaseAction) => {
    console.log(action);
};

export const defaultText = 'Hello World!';
