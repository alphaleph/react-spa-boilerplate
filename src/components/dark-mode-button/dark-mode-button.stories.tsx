import React from 'react';
import { PureDarkModeButton } from './dark-mode-button';
import { BaseAction } from '../../types';
import '../../index.scss';
import './dark-mode-button.scss';

const darkModeButtonPropsFactory = (isDarkMode: boolean, text: string) => ({
    isDarkMode: isDarkMode,
    dispatch: (action: BaseAction) => {
        console.log(action);
    },
    text: text,
});

const defaultText = 'Hello World!';

export const defaultStory = () => (
    <PureDarkModeButton {...darkModeButtonPropsFactory(false, defaultText)} />
);

export const darkMode = () => (
    <PureDarkModeButton {...darkModeButtonPropsFactory(true, defaultText)} />
);

export default {
    title: 'DarkModeButton',
    component: PureDarkModeButton,
    includeStories: ['defaultStory', 'darkMode'],
};
