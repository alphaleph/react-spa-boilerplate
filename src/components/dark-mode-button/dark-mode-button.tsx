import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
    activateDarkMode,
    deactivateDarkMode,
} from '../../redux/actions/darkModeActions';
import { defaultDispatch } from '../../test-util';
import { BaseAction, AppState } from '../../types';
import './dark-mode-button.scss';

interface DarkModeButtonProps extends DarkModeButtonOwnProps {
    // Redux State
    isDarkMode?: boolean;
    // Redux Dispatch
    dispatch?: (action: BaseAction) => void;
}

interface DarkModeButtonOwnProps {
    text: string;
}

export const PureDarkModeButton = ({
    isDarkMode = false,
    dispatch = defaultDispatch,
    text,
}: DarkModeButtonProps) => {
    const toggleDarkMode = () => {
        isDarkMode
            ? dispatch(deactivateDarkMode())
            : dispatch(activateDarkMode());
    };

    return (
        <button
            className={isDarkMode ? 'is-dark' : 'is-light'}
            onClick={toggleDarkMode}
        >
            {text}
        </button>
    );
};

export const DarkModeButton = (props: DarkModeButtonOwnProps) => {
    const isDarkMode = useSelector((state: AppState) => {
        return state.darkMode.isDarkMode;
    });

    const dispatch = useDispatch();

    return (
        <PureDarkModeButton
            {...props}
            isDarkMode={isDarkMode}
            dispatch={dispatch}
        />
    );
};
