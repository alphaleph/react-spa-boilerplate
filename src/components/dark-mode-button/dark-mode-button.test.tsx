import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { PureDarkModeButton } from './dark-mode-button';
import { defaultText } from '../../test-util';

const testFn = jest.fn();

describe('DarkModeButton', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders without crashing', () => {
        const { getByText } = render(<PureDarkModeButton text={defaultText} />);
        expect(getByText(defaultText)).toBeInTheDocument();
    });

    it('renders correctly', () => {
        const { getByText } = render(<PureDarkModeButton text={defaultText} />);
        expect(getByText(defaultText)).toBeVisible();
    });

    it('dispatches on click', () => {
        const { getByText } = render(
            <PureDarkModeButton text={defaultText} dispatch={testFn()} />
        );

        userEvent.click(getByText(defaultText));

        expect(testFn).toBeCalledTimes(1);
    });

    it('has not changed', () => {
        const { asFragment } = render(
            <PureDarkModeButton text={defaultText} />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
