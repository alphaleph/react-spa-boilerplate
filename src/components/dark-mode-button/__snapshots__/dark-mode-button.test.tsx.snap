// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DarkModeButton has not changed 1`] = `
<DocumentFragment>
  <button
    class="is-light"
  >
    Hello World!
  </button>
</DocumentFragment>
`;
